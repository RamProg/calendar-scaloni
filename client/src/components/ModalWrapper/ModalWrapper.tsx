type ModalWrapperProps = {
  onClose: () => void;
  footer: React.ReactNode;
  children: React.ReactNode;
};

const ModalWrapper: React.FC<ModalWrapperProps> = ({
  children,
  footer,
  onClose,
}) => {
  return (
    <div className="fixed inset-0 z-10 flex items-center justify-center overflow-y-auto">
      <div className="fixed inset-0 transition-opacity" aria-hidden="true">
        <div
          className="absolute inset-0 bg-gray-500 opacity-75"
          onClick={onClose}
        />
      </div>
      <div className="inline-block overflow-hidden text-left align-bottom transition-all transform bg-gray-100 rounded-lg shadow-xl sm:my-8 sm:align-middle sm:max-w-lg sm:w-full">
        {children}
        <div className="px-4 py-3 bg-gray-100 sm:px-6 sm:flex sm:flex-row-reverse">
          {footer}
        </div>
      </div>
    </div>
  );
};

export default ModalWrapper;
