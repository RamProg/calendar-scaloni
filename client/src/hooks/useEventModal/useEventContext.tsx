import { createContext } from 'react';

export const EventModalContext = createContext({
  isOpen: false,
  eventData: {
    title: '',
    description: '',
    startDate: '',
    endDate: '',
  },
  editOrAdd: 'add',
  onSave: (id?: string) => {},
  openModal: (data?: any) => {},
  closeModal: () => {},
  updateEventData: (data: any) => {},
});
