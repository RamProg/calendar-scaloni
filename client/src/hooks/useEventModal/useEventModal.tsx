import { useContext } from 'react';
import { EventModalContext } from '@/src/contexts/EventModal/EventModalContext/EventModalContext';

export const useEventModal = () => useContext(EventModalContext);
