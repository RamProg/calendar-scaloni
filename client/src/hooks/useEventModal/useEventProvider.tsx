import { ReactNode, useState } from 'react';
import { EventModalContext } from './useEventContext';
import { EventType } from '@/src/types';

const emptyEventData = {
  title: '',
  description: '',
  startDate: '',
  endDate: '',
};

export const EventModalProvider = ({ children }: { children: ReactNode }) => {
  const [isOpen, setIsOpen] = useState(false);
  const [editOrAdd, setEditOrAdd] = useState('add');
  const [eventData, setEventData] = useState(emptyEventData);

  const onSave = (id?: string) => {
    if (id) {
      // update event
    } else {
      // create event
    }
  };

  const updateEventData = (data: Partial<EventType>) => {
    setEventData((prev) => ({ ...prev, ...data }));
  };

  const openModal = (data?: EventType) => {
    if (!data) {
      setEditOrAdd('add');
      setEventData(emptyEventData);
    } else {
      setEditOrAdd('edit');
      updateEventData(data);
    }
    setIsOpen(true);
  };

  const closeModal = () => {
    setIsOpen(false);
  };

  return (
    <EventModalContext.Provider
      value={{
        isOpen,
        eventData,
        openModal,
        closeModal,
        onSave,
        updateEventData,
        editOrAdd,
      }}
    >
      {children}
    </EventModalContext.Provider>
  );
};
