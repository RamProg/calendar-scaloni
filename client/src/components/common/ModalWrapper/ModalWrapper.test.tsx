import '@testing-library/jest-dom';
import { render, fireEvent, screen } from '@testing-library/react';
import ModalWrapper from './ModalWrapper';
import { useEventModal } from '@/src/hooks/useEventModal/useEventModal';

jest.mock('@/src/hooks/useEventModal/useEventModal', () => ({
  useEventModal: jest.fn().mockReturnValue({
    serverErrors: { sendData: false },
  }),
}));

describe('ModalWrapper', () => {
  const mockOnClose = jest.fn();
  describe('props', () => {
    beforeEach(() => {
      render(
        <ModalWrapper onClose={mockOnClose} footer={<div>Footer content</div>}>
          <p>Modal content</p>
        </ModalWrapper>
      );
    });
    it('renders correctly', () => {
      const modalElement = screen.getByText('Modal content');
      expect(modalElement).toBeInTheDocument();
      expect(modalElement).toMatchSnapshot();
    });

    it('calls onClose prop when modal overlay is clicked', () => {
      const modalOverlay = screen.getByTestId(/modal-overlay/i);
      fireEvent.click(modalOverlay);
      expect(mockOnClose).toHaveBeenCalledTimes(1);
    });

    it('renders footer correctly', () => {
      const footerElement = screen.getByText('Footer content');
      expect(footerElement).toBeInTheDocument();
    });
  });

  describe('error states', () => {
    afterEach(() => {
      jest.clearAllMocks();
    });

    it('renders error message when serverErrors.sendData is true', () => {
      // Mock the serverErrors.sendData condition
      (useEventModal as jest.Mock).mockReturnValue({
        serverErrors: { sendData: true },
      });

      render(
        <ModalWrapper onClose={mockOnClose} footer={<div>Footer content</div>}>
          <p>Modal content</p>
        </ModalWrapper>
      );

      const errorMessage = screen.getByText(/Oops.. please try again later/i);
      expect(errorMessage).toBeInTheDocument();
    });

    it('does not render error message when serverErrors.sendData is false', () => {
      // Mock the serverErrors.sendData condition
      (useEventModal as jest.Mock).mockReturnValue({
        serverErrors: { sendData: false },
      });

      render(
        <ModalWrapper onClose={mockOnClose} footer={<div>Footer content</div>}>
          <p>Modal content</p>
        </ModalWrapper>
      );

      const errorMessage = screen.queryByText('Oops.. please try again later');
      expect(errorMessage).not.toBeInTheDocument();
    });
  });
});
