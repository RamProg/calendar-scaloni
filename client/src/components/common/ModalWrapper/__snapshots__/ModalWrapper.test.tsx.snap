// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalWrapper props renders correctly 1`] = `
<p>
  Modal content
</p>
`;
